add BLAKE3 hash

divinity76 opened the pull request:
BLAKE3 is a very fast cryptograpically secure hash. It is the latest iteration of the BLAKE hash, which was a SHA3 finalist (but lost to Keccak in the final round, for being too similar to SHA2).

Check this speed chart: 
![](https://raw.githubusercontent.com/BLAKE3-team/BLAKE3/master/media/speed.svg)

and this bench.php run where BLAKE3 is much faster than every secure hash, and beats several non-cryptographic hashes: 
```sh
$ sapi/cli/php ext/hash/bench.php
crc32b       0.001195
crc32c       0.001202
crc32        0.001202
xxh3         0.001234
xxh128       0.001289
xxh64        0.001475
xxh32        0.002235
murmur3f     0.002459
murmur3c     0.003681
murmur3a     0.004289
adler32      0.007718
blake3       0.007752
md4          0.013109
fnv132       0.015075
fnv164       0.015109
fnv1a64      0.015116
fnv1a32      0.015251
joaat        0.018858
md5          0.019797
sha1         0.020472
tiger160,3   0.021290
tiger192,3   0.021363
tiger128,3   0.021366
tiger128,4   0.027518
tiger160,4   0.027743
tiger192,4   0.027870
ripemd128    0.029190
ripemd256    0.029378
sha3-224     0.029787
sha3-256     0.031518
haval256,3   0.038328
haval224,3   0.038479
haval128,3   0.038483
sha3-384     0.038559
haval192,3   0.038564
haval160,3   0.039068
sha512/256   0.039302
sha512       0.039307
sha512/224   0.039472
sha384       0.039508
ripemd160    0.042287
ripemd320    0.043036
sha3-512     0.054044
haval192,4   0.055699
haval224,4   0.055902
haval160,4   0.055925
haval256,4   0.055948
haval128,4   0.056165
sha256       0.057846
sha224       0.058139
haval128,5   0.070442
haval224,5   0.070503
haval256,5   0.070569
haval192,5   0.070576
haval160,5   0.071109
whirlpool    0.086256
gost         0.200251
gost-crypto  0.200709
snefru256    0.449650
snefru       0.451111
md2          1.237880
```

- AFAIK the PHP project doesn't like git submodules (for example Dmitry Stogov wanted the new JIT engine to be a submodule but the PHP developers refused), so instead I made ext/hash/blake3/fetch_upstream_blake3.sh
- I have not added SSE2/AVX2/AVX512/etc optimized builds to MSVC because I don't have a MSVC system to test on, so I just added the bare-minimum portable implementation to MSVC builds. (feel free to fix it, I don't want to.)

mvorisek commented on ext/hash/blake3/fetch_upstream_blake3.sh:
```diff
@@ -0,0 +1,13 @@
+#!/bin/sh
```
this file should be added to https://github.com/php/php-src/blob/php-8.3.2/.github/actions/verify-generated-files/action.yml to assert the upstream files are not unintentionally modified

divinity76 replied:
@mvorisek like, just add it to the end of the script?
```
        build/gen_stub.php -f
        build/gen_stub.php --generate-optimizer-info
        git add . -N && git diff --exit-code
+      ext/hash/blake3/fetch_upstream_blake3.sh
```
like that?

divinity76 replied:
added: fcc16d318a1eebab7938c1421b7df3e92f0e5198

iluuu1994 replied:
That does not seem necessary. We don't do that for any other dependencies.

mvorisek replied:
@iluuu1994 I consider this highly useful and even as security hardening, with large bundled resources, every bundling/updating needs to be checked if the code matches 1:1 the original trusted sources (plus our possible patches). I hope we can agree comparing the sources manually is not better than this done by CI.

I think you are concerned by the extra CPU resources consumed by every push. I have an idea - let's add "verifyBundledBlake3" option to https://github.com/php/php-src/blob/master/.github/actions/verify-generated-files/action.yml which wil default to "false" and set to "true" only when the bundled sources were changed in "PR" or "last commit" or nightly run (to cover push with multiple commits).

This way, we can assert all other bundled sources like JIT IR, Zend/asm, ...

I am happy to your feedback, I hope you find this useful.

divinity76 replied:
IMO @mvorisek has a good point: it would be nice if we had a way to verify the integrity of large bundled resources. Even tho [my attempt](https://github.com/php/php-src/pull/13194#discussion_r1460496852) turned out to be controversial.

devnexen commented:
That s quite an addition ... worth discussing.

iluuu1994 commented:
This is indeed a lot of code. I think this should be discussed on the list, and possibly voted on.

divinity76 commented:
asked on php-internals mailing list: https://marc.info/?l=php-internals&m=170568974400302&w=2
(and fixed MSVC and ARM+Neon builds, the latter tested on Oracle Cloud ARM64)

divinity76 commented on ext/hash/php_hash_blake3.h:
```diff
@@ -0,0 +1,35 @@
+#define PHP_BLAKE3_CTX blake3_hasher
+// help: is V correct?
+//#define PHP_BLAKE3_SPEC "b8b8qb64bbbbb1760"
+#define PHP_BLAKE3_SPEC "L8L8Qa64CCCCL8Ca1760"
```
can someone double-check this? because i couldn't wrap my head around it, and this is just my best guess

divinity76 replied:
Look at line 150, uppercase Q still exists: https://github.com/php/php-src/blob/6d4598eba851b24ce537cc9390c140a61f661108/ext/hash/hash.c#L150

Pretty sure uppercase C also existed with the meaning `unsigned char` or `uint8_t` when it was written, but it seems to have been removed! I'll try to re-write it again.. thanks!

divinity76 replied:
Still can't get it quite right, do you see anything off here?
```c
#define PHP_BLAKE3_SPEC /* uint32_t key[8]; */"l8" \
/* uint32_t cv[8]; */ "l8" \
/* uint64_t chunk_counter; */ "q1" \
/* uint8_t buf[BLAKE3_BLOCK_LEN];  */ "b64" \
/* uint8_t buf_len; */ "b1" \
/* uint8_t blocks_compressed */ "b1" \
/* uint8_t flags; */ "b1" \
/* uint8_t cv_stack_len; */ "b1" \
/* uint8_t cv_stack[(BLAKE3_MAX_DEPTH + 1) * BLAKE3_OUT_LEN]; */ "b1760" \
"."
```

divinity76 replied:
I found a possible solution: it seems the compiler insert 5 alignment-padding-bytes after `uint8_t buf_len;`,
and an additional 7 padding bytes after `uint8_t cv_stack[(BLAKE3_MAX_DEPTH + 1) * BLAKE3_OUT_LEN];`, if the spec is changed to
```
#define PHP_BLAKE3_SPEC /* uint32_t key[8]; */"l8" \
/* uint32_t cv[8]; */ "l8" \
/* uint64_t chunk_counter; */ "q" \
/* uint8_t buf[BLAKE3_BLOCK_LEN];  */ "b64" \
/* uint8_t buf_len; */ "b" \
/* skip 5 bytes of alignment padding in chunk */ "B5" \
/* uint8_t blocks_compressed */ "b" \
/* uint8_t flags; */ "b" \
/* uint8_t cv_stack_len; */ "b" \
/* uint8_t cv_stack[(BLAKE3_MAX_DEPTH + 1) * BLAKE3_OUT_LEN]; */ "b1760" \
    /* skip 7 trailing alignment bytes */     "B7"   /* offset 1905..1911 -> 1912 */ \
"."
```
serialization actually works on my system!

But this seems like the kind of thing different compilers, and even the same compiler with different optimization settings, may choose to do differently. For example, I wouldn't trust `gcc -Ofast` (optimize-for-speed) and `gcc -Os` (optimize-for-size) to do the exact same alignments paddings in the exact same location 🤔 

I might be wrong but.. idk!

divinity76 replied:
should I just push the spec that "works on my system" to the branch and see what happens? I'd be interested to see if the 32bit CI tests pass or not 🤔

bukka replied:
you could maybe add one uint8_t reserved field and then on 64bit reserved2 field of uint32_t type which should hopefully make it defined on all platforms.

divinity76 replied:
tested replacing the default `-O2` with `-Os` locally, and the padding was still identical between O2 and Os at least.

bukka replied:
I think compiler is free to add padding but it cannot reorder fields so if the field is there, it needs to keep it that way and if it's padded to 64bits (on 64bit arch) or to 32bits (on 32bit arch), it shouldn't add any extra padding so I think it's better to add the reserved fields as I suggested.

divinity76 replied:
attempted fix: https://github.com/php/php-src/pull/13194/commits/8762e329d4da9460e9365087eda625eaa4380e41

divinity76 replied:
> you could maybe add one uint8_t reserved field and then on 64bit reserved2 field of uint32_t type which should hopefully make it defined on all platforms.

somehow didn't see that message  until after pushing 8762e329d4da9460e9365087eda625eaa4380e41 🤔 but I like that better. I'll run some tests locally

bukka replied:
From what I see, there's also option to define your own serialization function but not sure if it would make it any simpler.

bukka replied:
Well it could be cleaner because it wouldn't rely on struct memory layout...

divinity76 replied:
>I think 5 bytes is correct only on 64bit but it should not be correct on 32bit

well https://github.com/php/php-src/pull/13194/commits/263b2f6c217be0b3cf877ba86c43eb7aaf9a20a4 works on 32bit (tested locally on a debian 12 32bit docker image, and now confirmed on the 32bit github CI runner)

>From what I see, there's also option to define your own serialization function but not sure if it would make it any simpler.

hmm could you link to an example? I haven't seen it. A custom serializer sounds like the safest bet across compilers/architectures/configurations

bukka replied:
> well 263b2f6 works on 32bit (tested locally on a debian 12 32bit docker image, and now confirmed on the 32bit github CI runner)

so it pads to 64bit on 32bit? interesting. I thought that it gets padded to 32bit on 32bit arch but probably I'm wrong.

> hmm could you link to an example?

You can define your own php_hash_serialize_func_t and php_hash_unserialize_func_t that you can set to php_hash_ops instead of php_hash_serialize / php_hash_unserialize. Then you should be able to do a custom serialization / unserialization and ignore the spec.

devnexen commented on ext/hash/blake3/fetch_upstream_blake3.sh:
```diff
@@ -0,0 +1,13 @@
+#!/bin/sh
```
just a personal opinion, but pcre2 nor opcache's JIT code are cloned. I do not see a particular reason to make an exception.

divinity76 replied:
@devnexen this script makes it easy to see exactly how blake3 1.5.0 was cloned, and should make it easy to update to future versions (like 1.6.0) in the future - IMO it's worthwhile to keep this script. idk if it should be part of `.github/actions/verify-generated-files/action.yml` tho (there's a discussion on that topic here https://github.com/php/php-src/pull/13194#discussion_r1458860073 )

devnexen replied:
I do not see the appeal nor the necessity. Couple of counter arguments from the top of my head :
- does not make build easier, CI included.
- if this blake3 release contains bugs, the upgrade can be done in the same way the internal pcre2 library does. There is no necessity to upgrade versions if it brings nothing to php or goes against php's needs otherwise.
- if this contains bugs we are able to fix quickly but would take a long time (or worse getting refused) upstream, in your case we would need to patch it on the fly or forking the repo.

iluuu1994 replied:
@divinity76 The point of `verify-generated-files` is to avoid sneaking malicious code into generated files that are collapsed during code reviews that people don't generally look at. This risk doesn't exist here. IMO, if the intent is to avoid people mistakenly editing these files, the same issue already exists for all other copied projects, and should be solved universally (if at all).

divinity76 replied:
@iluuu1994 
>The point of verify-generated-files is to avoid sneaking malicious code into generated files that are collapsed during code reviews that people don't generally look at. This risk doesn't exist here. 

Interesting, thanks for the explanation! 

Maybe I snuck some malicious code into one of the upstream_blake3/* files? If I did, this script would actually remove it, and then `verify-generated-files` would notice it!

@devnexen 

>I do not see the appeal 

My main appeal is that if I want to update the bundled blake3 in the future, i can just change the 1 line `git clone --branch '1.5.0` with `git clone --branch '1.5.1'` and the script should do the rest.

>does not make build easier, CI included.

fair

>if this blake3 release contains bugs, the upgrade can be done in the same way the internal pcre2 library does.

How does upstream pcre2 fixes get applied to the internal pcre2 library? i don't know

>There is no necessity to upgrade versions if it brings nothing to php or goes against php's needs otherwise.

There's good updates sometimes, for example the latest v1.5.0 fixed a arm big-endian compile issue, v1.4.1 increased ARM performance by ~10-30%, 1.3.3 fixed a GCC6.1 AVX512 debug corruption issue.

>if this contains bugs we are able to fix quickly but would take a long time (or worse getting refused) upstream, in your case we would need to patch it on the fly or forking the repo.

Yeah that's a good point. If we get to that point, this script is useless. Anecdotal evidence, the only time I made a PR to BLAKE3, it was accepted in 11 hours ( https://github.com/BLAKE3-team/BLAKE3/pull/131 ), but it was a trivial one.

devnexen replied:
> There's good updates sometimes, for example the latest v1.5.0 fixed a arm big-endian compile issue, v1.4.1 increased ARM performance by ~10-30%, 1.3.3 fixed a GCC6.1 AVX512 debug corruption issue.

Those are valid reasons to upgrade but again upgrades can be handled in same manner at the aforementioned php components. Note that I m not against having blake3 feature, I see it as a "nice to have", just the way of the dependency being handled.

divinity76 replied:
removed the script.

divinity76 commented:
post on the mailing list worth re-posting here:
GCC11.4, even with -march=native -mtune=native, which is *not* commonly used in PHP,
the compiler didn't stand a chance against the hand-optimized assembly versions
![image](https://github.com/php/php-src/assets/1874996/c759ac1e-530b-403e-b312-6ddd68508902)


wrote some benchmarks, but the TL;DR is:
portable -O2 usually used by PHP managed 1126MB/s,
portable -O2 -march=native managed 533MB/s (wtf? gcc obviously got
something wrong here),
hand-written -O2 SSE2  managed 3144MB/s,
hand-written -O2 SSE41 managed 3332MB/s,
hand-written -O2 avx2 managed 6554MB/s,
hand-writen -O2 AVX512 managed 8913MB/s,
on my AMD Ryzen 9 7950x,
benchmarking code:
https://gist.github.com/divinity76/5729472dd5d77e94cd0acb245aac2226
raw output:
```php
array(6) {
  ["O2-portable-march"]=>
  array(2) {
    ["microseconds_for_16_kib"]=>
    int(29295)
    ["mb_per_second"]=>
    float(533.3674688513398)
  }
  ["O2-portable"]=>
  array(2) {
    ["microseconds_for_16_kib"]=>
    int(13876)
    ["mb_per_second"]=>
    float(1126.0449697319111)
  }
  ["O2-sse2"]=>
  array(2) {
    ["microseconds_for_16_kib"]=>
    int(4969)
    ["mb_per_second"]=>
    float(3144.4958744214127)
  }
  ["O2-sse41"]=>
  array(2) {
    ["microseconds_for_16_kib"]=>
    int(4688)
    ["mb_per_second"]=>
    float(3332.977815699659)
  }
  ["O2-avx2"]=>
  array(2) {
    ["microseconds_for_16_kib"]=>
    int(2384)
    ["mb_per_second"]=>
    float(6554.1107382550335)
  }
  ["O2-avx512"]=>
  array(2) {
    ["microseconds_for_16_kib"]=>
    int(1753)
    ["mb_per_second"]=>
    float(8913.291500285226)
  }
}
```

petk commented on ext/hash/config.m4:
```diff
@@ -17,31 +26,55 @@ else
   php_hash_whirlpool.h php_hash_adler32.h php_hash_crc32.h \
   php_hash_fnv.h php_hash_joaat.h php_hash_sha3.h php_hash_murmur.h \
-  php_hash_xxhash.h xxhash/xxhash.h"
+  php_hash_xxhash.h xxhash/xxhash.h php_hash_blake3.h"
```
```suggestion
  php_hash_xxhash.h xxhash/xxhash.h php_hash_blake3.h blake3/upstream_blake3/c/blake3.h"
```

To have the dependency header also installed when using the php_hash_blake3.h.

petk commented on ext/hash/config.w32:
```diff
@@ -40,4 +44,4 @@ PHP_INSTALL_HEADERS('ext/hash',	'php_hash.h php_hash_md.h php_hash_sha.h ' +
 									'php_hash_adler32.h php_hash_crc32.h php_hash_sha3.h ' +
 									'php_hash_murmur.h php_hash_xxhash.h php_hash_fnv.h ' +
-									'php_hash_joaat.h xxhash/xxhash.h');
+									'php_hash_joaat.h xxhash/xxhash.h php_hash_blake3.h');
```
```suggestion
									'php_hash_joaat.h xxhash/xxhash.h php_hash_blake3.h blake3/upstream_blake3/c/blake3.h');
```

petk commented on ext/hash/blake3/fetch_upstream_blake3.sh:
```diff
@@ -0,0 +1,13 @@
+rm -rf .git
+cd c
+# some stuff we don't need
+rm -rf blake3_c_rust_bindings test.py example.c main.c Makefile.testing CMakeLists.txt blake3-config.cmake.in README.md .gitignore
```
Sorry for partial review only, but maybe also the unused libblake3.pc.in can be removed here?

```suggestion
rm -rf blake3_c_rust_bindings test.py example.c main.c Makefile.testing CMakeLists.txt blake3-config.cmake.in README.md .gitignore libblake3.pc.in
```

bukka commented:
It's missing license in each file so that will need to be added.

divinity76 commented:
@bukka added the LICENSE file from https://github.com/BLAKE3-team/BLAKE3/blob/master/LICENSE , 
but adding them to every file sounds excessive, are you sure we actually want that?

petk commented:
Regarding the license, can we maybe add it to the README.REDIST.BINS file? Number `21. Blake3 (ext/hash/blake3)` and content of LICENSE file copied into it at the end. This file is also distributed when download PHP source code either for Windows or the tarball from [downloads.php.net](https://www.php.net/downloads.php) so it's on one place, and code is a bit less cluttered.

bukka commented:
I think the added LICENSE file and a note in README.REDIST.BINS should be enough for upstream files. You should add a header to the files that you added - hash_blake3.c and php_hash_blake3.h

petk commented on ext/hash/config.m4:
```diff
@@ -17,31 +26,55 @@ else
     SHA3_OPT_SRC="$SHA3_DIR/KeccakP-1600-inplace32BI.c"
   ],[
     AC_MSG_RESULT([yes])
+    if test $ac_cv_target_x86 = yes; then 
+      if test $ac_cv_target_windows = yes; then
+        dnl x86_64 windows gnuc
+        EXT_HASH_BLAKE3_SOURCES="$EXT_HASH_BLAKE3_SOURCES blake3/upstream_blake3/c/blake3_avx512_x86-64_windows_gnu.S blake3/upstream_blake3/c/blake3_avx2_x86-64_windows_gnu.S blake3/upstream_blake3/c/blake3_sse41_x86-64_windows_gnu.S blake3/upstream_blake3/c/blake3_sse2_x86-64_windows_gnu.S"
+      else
+        if test $ac_cv_target_unix = yes; then 
+          dnl x86_64 unix gnuc
+          EXT_HASH_BLAKE3_SOURCES="$EXT_HASH_BLAKE3_SOURCES blake3/upstream_blake3/c/blake3_avx512_x86-64_unix.S blake3/upstream_blake3/c/blake3_avx2_x86-64_unix.S blake3/upstream_blake3/c/blake3_sse41_x86-64_unix.S blake3/upstream_blake3/c/blake3_sse2_x86-64_unix.S"
+        else
+          dnl blake still has C instrictics versions (eg ext/hash/blake3/upstream_blake3/c/blake3_avx512.c  ) which COULD be used in this situation, 
+          dnl we're targeting x86_64 but neither windows nor unix.. 
+          dnl i have nothing to test this on, so i'm just using portable for now, feel free to fix it.
+          PHP_HASH_CFLAGS="$PHP_HASH_CFLAGS -DBLAKE3_NO_SSE2 -DBLAKE3_NO_SSE41 -DBLAKE3_NO_AVX2 -DBLAKE3_NO_AVX512"
+        fi
+      fi
+    fi
```
```suggestion
    AS_CASE([$host_cpu],[x86_64*|amd64*], [
      if test $ac_cv_target_windows = yes; then
        dnl x86_64 windows gnuc
        EXT_HASH_BLAKE3_SOURCES="$EXT_HASH_BLAKE3_SOURCES blake3/upstream_blake3/c/blake3_avx512_x86-64_windows_gnu.S blake3/upstream_blake3/c/blake3_avx2_x86-64_windows_gnu.S blake3/upstream_blake3/c/blake3_sse41_x86-64_windows_gnu.S blake3/upstream_blake3/c/blake3_sse2_x86-64_windows_gnu.S"
      else
        if test $ac_cv_target_unix = yes; then 
          dnl x86_64 unix gnuc
          EXT_HASH_BLAKE3_SOURCES="$EXT_HASH_BLAKE3_SOURCES blake3/upstream_blake3/c/blake3_avx512_x86-64_unix.S blake3/upstream_blake3/c/blake3_avx2_x86-64_unix.S blake3/upstream_blake3/c/blake3_sse41_x86-64_unix.S blake3/upstream_blake3/c/blake3_sse2_x86-64_unix.S"
        else
          dnl blake still has C instrictics versions (eg ext/hash/blake3/upstream_blake3/c/blake3_avx512.c  ) which COULD be used in this situation, 
          dnl we're targeting x86_64 but neither windows nor unix.. 
          dnl i have nothing to test this on, so i'm just using portable for now, feel free to fix it.
          PHP_HASH_CFLAGS="$PHP_HASH_CFLAGS -DBLAKE3_NO_SSE2 -DBLAKE3_NO_SSE41 -DBLAKE3_NO_AVX2 -DBLAKE3_NO_AVX512"
        fi
      fi
    ])
```

This detects the x86_64 processor. x86_64 is detected as word amd64 on FreeBSD. The $host_os is in Autoconf used for the target CPU when cross-compiling or as a host CPU when building on the same machine as the program will be also run.

petk commented on ext/hash/config.m4:
```diff
@@ -17,31 +26,55 @@ else
   AC_MSG_CHECKING([if we're at 64-bit platform])
   AS_IF([test "$ac_cv_sizeof_long" -eq 4],[
     AC_MSG_RESULT([no])
+    if test $ac_cv_target_x86 = yes; then 
+      dnl there are some early Pentium4 32bit CPUs with SSE2 support, but don't waste effort optimizing musem-CPUs
+      dnl (the majority of 32bit CPUs does not have SSE2)
+      PHP_HASH_CFLAGS="$PHP_HASH_CFLAGS -DBLAKE3_NO_SSE2 -DBLAKE3_NO_SSE41 -DBLAKE3_NO_AVX2 -DBLAKE3_NO_AVX512"
+    fi
```
```suggestion
    AS_CASE([$host_cpu],[x86*|amd*|i?86*|pentium],[
      dnl there are some early Pentium4 32bit CPUs with SSE2 support, but don't waste effort optimizing musem-CPUs
      dnl (the majority of 32bit CPUs does not have SSE2)
      PHP_HASH_CFLAGS="$PHP_HASH_CFLAGS -DBLAKE3_NO_SSE2 -DBLAKE3_NO_SSE41 -DBLAKE3_NO_AVX2 -DBLAKE3_NO_AVX512"
])
```

This detects x86 32-bit processor (can be x86, i386 or similar, or "amd" on FreeBSD.

divinity76 commented:
@bukka @petk added LICENSE headers and reference: 1954b37c3933e4992dd8baf0712fbd4804739075 - I prefer a reference in README.REDIST.BINS rather than in-lining it.

cypherbits commented:
Having this would be good

cypherbits commented:
More than 2 years since the first attempt to get Blake3 into PHP, why is it so hard?

divinity76 commented:
> More than 2 years since the first attempt to get Blake3 into PHP, why is it so hard?

the main problem this round was that I just lost steam. Got occupied with other stuff and forgot about this

vodnicearv commented:
BLAKE3 is very nice, we need that algorithm in php

leonardocustodio commented:
Seems like this will never be merged, such a shame as a lot of blockchain are starting to use blake3 as standard

divinity76 commented:
Fwiw this should be rebased against the newest blake3 release, iirc (on phone now can't check, but iirc) this pr currently contains 3 patches to blake3 with a dedicated patch file, but all 3 patches have been accepted upstream and is part of the newest Blake3 official release

divinity76 commented:
Scratch that, https://github.com/BLAKE3-team/BLAKE3/pull/382 is still not resolved, 
so if we want to compile php cleanly on MacOS, we still need at least 1 patch for blake3.

the other 2 are resolved upstream tho.

divinity76 commented:
- removed [controversial](https://github.com/php/php-src/pull/13194#discussion_r1460496852) fetch_upstream_blake3.sh/patches.diff/verify-generated-files stuff.
- updated branch to php8.5-dev master.
- updated to upstream blake3 1.5.5
- added some build fixes -Wall -Werror fixes: https://github.com/BLAKE3-team/BLAKE3/pull/382 / https://github.com/BLAKE3-team/BLAKE3/pull/443
- i think the [MACOS_X64_DEBUG_NTS](https://github.com/php/php-src/actions/runs/12916615779/job/36021168169?pr=13194#logs) failure is not my fault, it appears to be a random timeout error.
(PS would be nice if someone could double-check that PHP_BLAKE3_SPEC looks correct.)

I'm requesting re-review.

remicollet commented:
-1 from me

I think it is a terrible idea to bundle more crypto stuff in PHP
For such serious things, we should rely on external libraries (OpenSSL, sodium...) provided by project specialized in crypto.
We already have bundled far too much things.

Is there an RFC for this ?

divinity76 commented:
@remicollet imo we should support BLAKE3 for the same reason we support xxHash (added 8.1.0): speed. BLAKE3 offers SHA3-256-like security at much higher speed than SHA3-256. Quoting `/ext/hash/bench.php`: sha3-256 0.031518 blake3 0.007752.
![output (2)](https://github.com/user-attachments/assets/6bd7e8f9-3a8b-4903-9a1a-a0ca735a1477)


Exactly how we support BLAKE3 is not that important, if OpenSSL/Sodium starts offering BLAKE3, I wouldn't mind requiring OpenSSL/Sodium for BLAKE3 support. But they don't support BLAKE3 (yet?).


As for RFC, made a draft a year ago, but now I can't even find it. I'll make a new one.

remicollet commented:
> Exactly how we support BLAKE3 is not that important, if OpenSSL/Sodium starts offering BLAKE3, I wouldn't mind requiring OpenSSL/Sodium for BLAKE3 support. But they don't support BLAKE3 (yet?).

So effort should go there, not here.

divinity76 commented:
RFC draft: https://wiki.php.net/rfc/blake3

bukka commented on ext/hash/config.m4:
```diff
@@ -73,6 +106,8 @@ PHP_INSTALL_HEADERS([ext/hash], m4_normalize([
   php_hash_sha.h
   php_hash_sha3.h
+  php_hash_blake3.h
+  blake3/upstream_blake3/c/blake3.h
```
Why is there that extra upstream_blake3 directory. Cannot this go directly to blake3?

divinity76 replied:
the directory separate code written for the php-src integration from the substantial amount of code cloned directly from https://github.com/BLAKE3-team/BLAKE3/

bukka replied:
I don't really understand why it needs extra subdirectory. The blake3 directly contains only upstream_blake3 directory which really doesn't make any sense to me. And you also have c directory which I can see is from upstream. So it should really be just blake3/c .

bukka commented on ext/hash/tests/hash-clone.phpt:
```diff
@@ -10,9 +10,13 @@ foreach ($algos as $algo) {
+    $orig_final = hash_final($orig);
+    $copy_final = hash_final($copy);
+    var_dump($orig_final);
+    if($orig_final !== $copy_final) {
```
NIT: space after if

bukka commented on ext/hash/tests/hash_copy_001.phpt:
```diff
@@ -10,9 +10,13 @@ foreach ($algos as $algo) {
+    $orig_final = hash_final($orig);
+    $copy_final = hash_final($copy);
+    var_dump($orig_final);
+    if($orig_final !== $copy_final) {
```
NIT: space after if

bukka commented on ext/hash/tests/hash_hmac_algos.phpt:
```diff
@@ -37,6 +37,7 @@ Array
     [%d] => snefru256
     [%d] => gost
     [%d] => gost-crypto
+    [29] => blake3
```
```suggestion
    [%d] => blake3
```

bukka reviewed (commented):
Except the missing serialization, it seem ok to me. It would be also good to add a bit more tests to test some predefined vectors to be sure all is good.

I think it's good already for the RFC so you can announce it.

In terms of the maintanence, the hash extension bundles already quite a few of algos so this is just another one. It's taken from upstream like some other algorithms so I don't think it matters that much the amount of code. The upstream seems to be quite well maintained.

divinity76 commented on ext/hash/tests/hash_serialize_001.phpt:
```diff
@@ -41,7 +41,11 @@ foreach ($algos as $algo) {
+    return in_array($algo, [
+        "xxh3", 
+        "xxh128",
+        "blake3", // todo: blake3 can be seralized but it's not implemented yet
```
I suspect PHP_BLAKE3_SPEC is wrong, and that it breaks serialization, but I haven't been able to figure it out yet 🤔

divinity76 replied:
was a padding issue.

divinity76 commented:
Right now there is a problem with https://dev.mysql.com/ causing Windows CI's to fail, specifically https://dev.mysql.com/get/Downloads/MySQL-8.0/mysql-8.0.31-winx64.zip returns HTTP 403 Forbidden.

It's not this PR's fault 🤔

mvorisek commented:
PHP needs fast hash, can the RFC be moved to voting?

divinity76 commented:
@mvorisek last time I brought it up on the internals mailing list, they were worried about how the ext/hash API were getting cluttered, how it would increase compile time, and started discussing how less-frequently-used-hashes could be moved into pecl extensions (iirc) ... I dropped out of the conversation rather quickly 😅

mvorisek commented:
Hmm, I cannot comment on the compile times, but given this hash algo is currently the best one given its entropy, cryptographic safety and speed, I would say it should be integrated into the core and made available on every platform.

The maintanance should be also minimal given the source can be verified using CI and thus only the integration needs to be maintained mainly.

Given these, I belive there is enough rationale to integrate this PR. What about simply opening the voting with a possible optional question if the CI sources verification should be added as well. If the voting would pass, great, if not, we cannot do much other than pecl..

bukka commented:
It's hard to say if those comments would mean that the RFC cannot pass. I don't think it will be much more extra maintenance than other hashes as they are just taken from upstream. In this case the upstream is actually quite lively compare to others. The deciding factor should be also what's good for our users. So I would try to at least put it to vote and see the result then. I think it's worth a try for sure.

TimWolla commented:
@divinity76 It would certainly help if you'd update the benchmarks to include the latest SHA-2 changes. At least the SSE2 variant should be available everywhere. So it would probably make sense to list SHA-2 separately with the SSE2 implementation and the SHA-NI implementation that should be supported by your CPU.

TimWolla commented:
And looking back at the thread, you should also specify which of the Blake2 implementations is being used (SSE2, SSE4.1, AVX2, AVX512), because otherwise it's not an apples to apples comparison.

midnight-wonderer commented:
Is it possible to add tests for `_keyed`, and `_derive_key` variants too?  
I want to use BLAKE3 as a KDF.

Or this belongs to another RFC/PR?
